File: TimberCruisingApp/Persistence/Repositories/HeightDiameterFitRepository.swift
// Spec §8 + §7.4 (per project, per species). Rolling updates on plot close.

import Foundation
import CoreData
import Models
import InventoryEngine

public protocol HeightDiameterFitRepository {
    func create(_ f: HeightDiameterFit) throws -> HeightDiameterFit
    func read(id: UUID) throws -> HeightDiameterFit?
    func update(_ f: HeightDiameterFit) throws -> HeightDiameterFit
    func delete(id: UUID) throws
    func forProjectAndSpecies(projectId: UUID, speciesCode: String) throws -> HeightDiameterFit?
    func listByProject(_ projectId: UUID) throws -> [HeightDiameterFit]
}

public extension HeightDiameterFitRepository {

    /// §7.4 rolling update. Loads the existing fit for (project, species)
    /// if one exists, refits Näslund H–D using HDModel.update with the
    /// supplied observations as warm-start data, and either creates or
    /// updates the persisted row. Returns the persisted HeightDiameterFit.
    ///
    /// Callers (e.g. PlotClose) pass *all* measured (dbh, height) pairs
    /// for the species, not a delta — the fit is always over the full
    /// accumulated dataset.
    @discardableResult
    func recomputeForSpecies(
        projectId: UUID,
        speciesCode: String,
        observations: [(dbhCm: Float, heightM: Float)],
        minN: Int = 8,
        now: Date = Date()
    ) throws -> HeightDiameterFit {
        let prev = try forProjectAndSpecies(
            projectId: projectId, speciesCode: speciesCode)
        let warm: HDModel.Fit? = prev.flatMap { fit in
            HDModel.Fit.fromCoefficients(
                fit.coefficients, nObs: fit.nObs, rmse: fit.rmse)
        }
        let newFit = try HDModel.update(
            previous: warm,
            observations: observations,
            minN: minN)
        if let existing = prev {
            let updated = HeightDiameterFit(
                id: existing.id,
                projectId: projectId,
                speciesCode: speciesCode,
                modelForm: "naslund",
                coefficients: newFit.coefficients,
                nObs: newFit.nObs,
                rmse: newFit.rmse,
                updatedAt: now)
            return try update(updated)
        } else {
            let fresh = HeightDiameterFit(
                id: UUID(),
                projectId: projectId,
                speciesCode: speciesCode,
                modelForm: "naslund",
                coefficients: newFit.coefficients,
                nObs: newFit.nObs,
                rmse: newFit.rmse,
                updatedAt: now)
            return try create(fresh)
        }
    }
}

public final class CoreDataHeightDiameterFitRepository: HeightDiameterFitRepository {
    private let stack: CoreDataStack
    public init(stack: CoreDataStack) { self.stack = stack }

    public func create(_ f: HeightDiameterFit) throws -> HeightDiameterFit {
        try performWrite(stack: stack) { ctx in
            let e = try insert(HeightDiameterFitEntity.self,
                               entityName: "HeightDiameterFitEntity", in: ctx)
            HeightDiameterFitMapper.apply(f, to: e)
            return f
        }
    }

    public func read(id: UUID) throws -> HeightDiameterFit? {
        try performRead(stack: stack) { ctx in
            guard let e = try fetchOne(HeightDiameterFitEntity.self,
                                       entityName: "HeightDiameterFitEntity",
                                       keyPath: "id", value: id as CVarArg, in: ctx)
            else { return nil }
            return HeightDiameterFitMapper.toStruct(e)
        }
    }

    public func update(_ f: HeightDiameterFit) throws -> HeightDiameterFit {
        try performWrite(stack: stack) { ctx in
            guard let e = try fetchOne(HeightDiameterFitEntity.self,
                                       entityName: "HeightDiameterFitEntity",
                                       keyPath: "id", value: f.id as CVarArg, in: ctx)
            else { throw CoreDataError.notFound(id: f.id.uuidString) }
            HeightDiameterFitMapper.apply(f, to: e)
            return f
        }
    }

    public func delete(id: UUID) throws {
        try performWrite(stack: stack) { ctx in
            guard let e = try fetchOne(HeightDiameterFitEntity.self,
                                       entityName: "HeightDiameterFitEntity",
                                       keyPath: "id", value: id as CVarArg, in: ctx)
            else { throw CoreDataError.notFound(id: id.uuidString) }
            ctx.delete(e)
        }
    }

    public func forProjectAndSpecies(projectId: UUID, speciesCode: String) throws -> HeightDiameterFit? {
        try performRead(stack: stack) { ctx in
            let pred = NSPredicate(format: "projectId == %@ AND speciesCode == %@",
                                   projectId as CVarArg, speciesCode)
            let req = NSFetchRequest<HeightDiameterFitEntity>(entityName: "HeightDiameterFitEntity")
            req.predicate = pred
            req.sortDescriptors = [NSSortDescriptor(key: "updatedAt", ascending: false)]
            req.fetchLimit = 1
            guard let e = try ctx.fetch(req).first else { return nil }
            return HeightDiameterFitMapper.toStruct(e)
        }
    }

    public func listByProject(_ projectId: UUID) throws -> [HeightDiameterFit] {
        try performRead(stack: stack) { ctx in
            let pred = NSPredicate(format: "projectId == %@", projectId as CVarArg)
            let sort = [NSSortDescriptor(key: "speciesCode", ascending: true)]
            return try fetchMany(HeightDiameterFitEntity.self,
                                 entityName: "HeightDiameterFitEntity",
                                 predicate: pred, sort: sort, in: ctx)
                .map(HeightDiameterFitMapper.toStruct)
        }
    }
}

File: android/app/src/main/java/com/hcjeong/forestix/data/cruise/HeightDiameterFitRepository.kt
// Port of iOS Persistence/Repositories/HeightDiameterFitRepository.swift.
// Spec §8 + §7.4 (per project, per species). Rolling updates on plot close.
//
// The iOS protocol-extension `recomputeForSpecies` becomes a Kotlin
// extension function on the interface so every implementation inherits it,
// exactly like a Swift protocol extension.

package com.hcjeong.forestix.data.cruise

import com.hcjeong.forestix.inventory.HDModel
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.map
import java.util.UUID

interface HeightDiameterFitRepository {
    suspend fun create(f: HeightDiameterFit): HeightDiameterFit
    suspend fun read(id: UUID): HeightDiameterFit?
    suspend fun update(f: HeightDiameterFit): HeightDiameterFit
    suspend fun delete(id: UUID)
    suspend fun forProjectAndSpecies(projectId: UUID, speciesCode: String): HeightDiameterFit?
    suspend fun listByProject(projectId: UUID): List<HeightDiameterFit>

    /// Live view of `listByProject()` for Compose screens.
    fun observeByProject(projectId: UUID): Flow<List<HeightDiameterFit>>
}

/// §7.4 rolling update. Loads the existing fit for (project, species)
/// if one exists, refits Näslund H–D using HDModel.update with the
/// supplied observations as warm-start data, and either creates or
/// updates the persisted row. Returns the persisted HeightDiameterFit.
///
/// Callers (e.g. PlotClose) pass *all* measured (dbh, height) pairs
/// for the species, not a delta — the fit is always over the full
/// accumulated dataset.
///
/// Throws HDModel.FitError when the observations cannot support a fit.
suspend fun HeightDiameterFitRepository.recomputeForSpecies(
    projectId: UUID,
    speciesCode: String,
    observations: List<HDModel.Observation>,
    minN: Int = 8,
    now: Long = System.currentTimeMillis(),
): HeightDiameterFit {
    val prev = forProjectAndSpecies(projectId = projectId, speciesCode = speciesCode)
    val warm: HDModel.Fit? = prev?.let { fit ->
        HDModel.Fit.fromCoefficients(fit.coefficients, nObs = fit.nObs, rmse = fit.rmse)
    }
    val newFit = HDModel.update(
        previous = warm,
        observations = observations,
        minN = minN)
    return if (prev != null) {
        update(HeightDiameterFit(
            id = prev.id,
            projectId = projectId,
            speciesCode = speciesCode,
            modelForm = "naslund",
            coefficients = newFit.coefficients,
            nObs = newFit.nObs,
            rmse = newFit.rmse,
            updatedAt = now))
    } else {
        create(HeightDiameterFit(
            id = UUID.randomUUID(),
            projectId = projectId,
            speciesCode = speciesCode,
            modelForm = "naslund",
            coefficients = newFit.coefficients,
            nObs = newFit.nObs,
            rmse = newFit.rmse,
            updatedAt = now))
    }
}

class RoomHeightDiameterFitRepository(
    private val dao: HeightDiameterFitDao,
) : HeightDiameterFitRepository {

    override suspend fun create(f: HeightDiameterFit): HeightDiameterFit {
        dao.upsert(HeightDiameterFitMapper.apply(f))
        return f
    }

    override suspend fun read(id: UUID): HeightDiameterFit? =
        dao.byId(id)?.let(HeightDiameterFitMapper::toStruct)

    override suspend fun update(f: HeightDiameterFit): HeightDiameterFit {
        dao.byId(f.id) ?: throw CruiseDataError.NotFound(f.id.toString())
        dao.upsert(HeightDiameterFitMapper.apply(f))
        return f
    }

    override suspend fun delete(id: UUID) {
        if (dao.deleteById(id) == 0) throw CruiseDataError.NotFound(id.toString())
    }

    override suspend fun forProjectAndSpecies(
        projectId: UUID,
        speciesCode: String,
    ): HeightDiameterFit? =
        dao.forProjectAndSpecies(projectId, speciesCode)
            ?.let(HeightDiameterFitMapper::toStruct)

    override suspend fun listByProject(projectId: UUID): List<HeightDiameterFit> =
        dao.listByProject(projectId).map(HeightDiameterFitMapper::toStruct)

    override fun observeByProject(projectId: UUID): Flow<List<HeightDiameterFit>> =
        dao.observeByProject(projectId)
            .map { rows -> rows.map(HeightDiameterFitMapper::toStruct) }
}
